sdk/core/azure-core/src/main/java/com/azure/core/implementation/jackson/ObjectMapperShim.java

```diff
@@ -275,18 +283,40 @@ private JavaType createJavaType(Type type) {
                 javaTypeArguments[i] = createJavaType(actualTypeArguments[i]);
             }
 
-            return getFromCache(type, t -> mapper.getTypeFactory()
+            return getFromCache(TYPE_TO_JAVA_TYPE_CACHE, type, t -> mapper.getTypeFactory()
                 .constructParametricType((Class<?>) parameterizedType.getRawType(), javaTypeArguments));
         } else {
-            return getFromCache(type, t -> mapper.getTypeFactory().constructType(t));
+            return getFromCache(TYPE_TO_JAVA_TYPE_CACHE, type, t -> mapper.getTypeFactory().constructType(t));
         }
     }
 
+    @SuppressWarnings("unchecked")
     public <T> T deserialize(HttpHeaders headers, Type deserializedHeadersType) throws IOException {
         if (deserializedHeadersType == null) {
             return null;
         }
 
+        try {
+            Class<?> headersClass = TypeUtil.getRawClass(deserializedHeadersType);
+
+            MethodHandle constructor = getFromCache(TYPE_TO_STRONGLY_TYPED_HEADERS_CONSTRUCTOR_CACHE,
+                deserializedHeadersType, type -> {
+                    try {
+                        MethodHandles.Lookup lookup = ReflectionUtilsApi.INSTANCE.getLookupToUse(headersClass);
+                        return lookup.unreflectConstructor(headersClass.getDeclaredConstructor(HttpHeaders.class));
+                    } catch (Throwable throwable) {
+                        return null;
+                    }
+                });
+
+            if (constructor != null) {
+                return (T) constructor.invokeWithArguments(headers);
+            }
+        } catch (Throwable throwable) {
+            LOGGER.verbose("Failed to find or use MethodHandle Constructor that accepts HttpHeaders for "
```

Comment:
when something fails, would it be possible to cache the outcome and never try to do reflection above again for given deserializedHeadersType?

Reply:
Yes, that change can and should be made

Reply:
Turns out this already is done, if the MethodHandle lookup fails it will set a null value in the cache map. Then further down since the lookup is null it will fallback to previous behavior.